twister: make skip reason clearer

PerMac opened the pull request:
The aim of this PR is to make the skip reasons less obscure. Without it, there is a wide class of configurations getting "Not runnable on device" status, which is quite obscure. There are several different reasons for this status: harness not supported, fixture not available, missing installation of a simulator, etc. Statuses are clearer with this PR

Before:
`scripts/twister  -T samples/boards/nrf/clock_skew -T samples/basic/blinky -T tests/drivers/gpio/gpio_basic_api -v -v --inline-logs -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0 `:

DEBUG   - nrf52840dk_nrf52840       samples/boards/nrf/clock_skew/sample.boards.nrf.clock_skew SKIPPED: Not runnable on device
DEBUG   - nrf52840dk_nrf52840       samples/basic/blinky/sample.basic.blinky           SKIPPED: Not runnable on device
DEBUG   - nrf52840dk_nrf52840       tests/drivers/gpio/gpio_basic_api/drivers.gpio.2pin SKIPPED: Not runnable on device
DEBUG   - nrf52840dk_nrf52840       tests/drivers/gpio/gpio_basic_api/drivers.gpio.nrf_sense_edge SKIPPED: Not runnable on 

and 
`scripts/twister -T samples/hello_world -p nsim_hs -p hifive1 -v -v --filter runnable`:

DEBUG   - nsim_hs                   samples/hello_world/sample.basic.helloworld        SKIPPED: Not runnable on device
DEBUG   - hifive1                     samples/hello_world/sample.basic.helloworld        SKIPPED: Not runnable on device

With the PR the above commands produce:

DEBUG   - nrf52840dk_nrf52840       samples/boards/nrf/clock_skew/sample.boards.nrf.clock_skew SKIPPED: Not runnable on device: test is marked as build-only.
DEBUG   - nrf52840dk_nrf52840       samples/basic/blinky/sample.basic.blinky           SKIPPED: Not runnable on device: Harness led is not supported.
DEBUG   - nrf52840dk_nrf52840       tests/drivers/gpio/gpio_basic_api/drivers.gpio.2pin SKIPPED: Not runnable on device: Fixture gpio_loopback is not available.
DEBUG   - nrf52840dk_nrf52840       tests/drivers/gpio/gpio_basic_api/drivers.gpio.nrf_sense_edge SKIPPED: Not runnable on device: Fixture gpio_loopback is not available.

and

DEBUG   - nsim_hs                   samples/hello_world/sample.basic.helloworld        SKIPPED: Not runnable on device: nsimdrv not found.

DEBUG   - hifive1                     samples/hello_world/sample.basic.helloworld        SKIPPED: Not runnable on device: renode not found.

Each commit from this PR is explaining the changes introduced.

PerMac committed:
Twister: Add clear status for depends_on skip.

The skip reason was improved by clearly stating
which keyword is not supported.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac committed:
twister: remove duplicated code. Simplify logic for test "runnability".

There is no point in executing check_runnable() twice: when loading
test data from yaml and when filters are applied during test plan
creation. The check is left only in the latter case.
"tfilter" variable is removed since the same logic can be preserved
without it by directly using the "runnable" one.
Checks for harness and fixtures are encapsulated as individual methods.
Now checking fixtures both from cli and hwmap happens in a single
place. The check for runnability is tangled with apply_filter() method.
This commit makes the apply_filter() to be executed for all twisters
workflows to improve consitency of the results. Before, e.g. the
filters were not applied when --test-only was used. This would cause
errors if tests are first build with --build-only on one setup and
then ported to another setup with different configuration, where some
requirements (e.g. fixture) are not met.
The changes are followed with an update to twister unit tests to
reflect the modifications.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

nashif commented:
are you going to push an update to pass CI?

PerMac commented:
Yes. I've done quite some local testing with various scenarios and I hope I managed to get the logic right. I didn't check yet if the tests/samples failing in twister-build stage are PR related. There are for sure tests of twister itself (pytest) failing and they will require updating, since they use functions I modified and they have to be adapted to this modified logic.
 Unluckily, I won't be able to properly analyze at fix it soon, since I will be on holidays. I will come back to it in 2 weeks. Do you prefer closing this PR until I have time to get back to it?

nashif commented:
> Do you prefer closing this PR until I have time to get back to it?

no. I was just wondering if I should start revieweing or if there is an update coming to fix CI

PerMac commented:
I rebased the PR and added a WIP commit where I started to update twister tests. At least we will see if the failures in twister-build were/are PR related.

evgeniy-paltsev commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -656,8 +635,22 @@ def apply_filters(self, **kwargs):
+
+                if runnable:
+                    if not instance.run:
+                        instance.add_filter(f"Not runnable on device. {verdict}", Filters.PLATFORM)
```
Minor nitpick: probably we can rephrase is to:
```
"Not runnable on device because: {verdict}"
```

Currently it may look like two separate issues are reported, so it may me a bit confusing:
```
Not runnable on device. Harness led is not supported.
Not runnable on device. nsimdrv not found.
```

Instead it would be:
```
Not runnable on device because: Harness led is not supported.
Not runnable on device because: nsimdrv not found.
```

Feel free to ignore :)

PerMac replied:
Good point. Changed.

evgeniy-paltsev replied:
Hmm, probably you have forgotten to add it to PR - at least I don't see any changes the code related to this log message printing / generation.

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -160,8 +160,8 @@ def load(self):
-        else:
-            self.apply_filters()
+
+        self.apply_filters()
```
Not sure how all of this will work, twister gets a pre-deteremined and set testplan, why does not to apply filters again? I know there might be edge cases of running twister on one host and loading the plan somewhere else which might yield different results, but I think this can be addressed using other means, without changing how this workflow fundemntaly works. Try this:

```
./scripts/twister -e kernel -T tests/kernel/semaphore/
```
Nothing will be built or run, all good. Now put this in a test plan.

```
./scripts/twister -e kernel -T tests/kernel/sempahore --save-tests plan.json
```
then

```
./scripts/twister --load-tests plan.json
```

Tests will run, although the plan says otherwise.

nashif replied:
> Are you referring to these failures in clang build on non-native_posix platforms, even though they were not in the plan? Which was an issue with my PR, that I asked for help thinking they are CI based? Thanks for pointing me out. I believe it is fixed now. Like I said, now apply_filters() is not adding new instances if there were already existing ones.

no, I am not. 

Just try the above, a testplan which is not supposed to run anything will be completely ignored and tests will run because the original filters will be ignored if you apply the filter again without the actual filter being supplied on the command line.

PerMac replied:
You're right, I missed it... This happened since " it doesn't add new instances during apply_filters() if the instances were already existing. " However, if an instance was existing in the test plan it was evaluated whatever was its status. I agree it was incorrect. 
I fixed it to ignore the existing ones with `status=filtered`. 
I still think the workflow I am proposing is beneficial. As another advantage it allows tuning of the received test plan to limit it furtherly. E.g. `./scripts/twister  -T tests/kernel/semaphore/ --save-tests plan.json` and than `./scripts/twister --load-tests plan.json -e kernel`. Maybe the example is not the most useful one, but shows the idea. 

BTW I wonder if "filtered" tests should be in the test plan? Plan files can get very heavy, and filtered cases can burden it quit a lot. Maybe they should be omitted by default from the plan and stored in a separate file when `--store-filtered <path>` is used?

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -658,8 +640,22 @@ def apply_filters(self, **kwargs):
+
+                if runnable:
+                    if not instance.run:
+                        instance.add_filter(f"Not runnable on device: {verdict}", Filters.PLATFORM)
```
@evgeniy-paltsev here? I used fixup and rebased, so no new commit. Looks like you suggested with extra `because` but I guess changing `.` to `:` keeps the gist of your comment?

evgeniy-paltsev replied:
Ah, I see. Haven't notice the extra dot. :)
Thanks!

gopiotr commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -119,19 +119,39 @@ def get_case_or_create(self, name):
+        verdict = "NA"
+        fixture_required = self.testsuite.harness_config.get('fixture')
+
+        if not fixture_required:
```
To simplify logic please consider change this code to this:
```python
if fixture_required:
    ...
else:
   ...
```

gopiotr reviewed (commented):
Generally it's look okay for me. And I agree that in some cases it could be valuable to apply filters even if test instances are loaded from previously generated file. Like for example in case when we build all possible tests on one node, and then in another one we would like to run only specific tests which could be filtered by tag.

PerMac commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -119,19 +119,39 @@ def get_case_or_create(self, name):
+        verdict = "NA"
+        fixture_required = self.testsuite.harness_config.get('fixture')
+
+        if not fixture_required:
```
Done.

PerMac commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -182,45 +202,54 @@ def setup_handler(self, env):
     # Global testsuite parameters
-    def check_runnable(self, enable_slow=False, filter='buildable', fixtures=[]):
-
+    def check_runnable(self, enable_slow=False, runnable = False):
```
Good catch. Fixed (removed).

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -658,8 +640,22 @@ def apply_filters(self, **kwargs):
+                # instance.run tells twister if a test can be build and executed (if true) or only build (if false)
+                instance.run, verdict = instance.check_runnable(self.options.enable_slow, runnable)
+
+                harness_supported, verdict_harness = instance.is_harness_supported()
```
Good point. Moved them back to `check_runnable`. Now `instance.run` is only changed by `check_runnable`

PerMac commented:
@gopiotr Thanks for the comments, I applied your suggestions and force-pushed the PR.

gopiotr commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -631,21 +622,12 @@ def apply_filters(self, **kwargs):
+                    if instance.name not in self.instances:
+                        continue
+                    elif self.instances[instance.name].status == "filtered":
+                        continue
```
`run_id` of new created `TestInstance` should be inherited from old one. For example:

```python
else:
    instance.run_id = self.instances[instance.name].run_id
```

PerMac replied:
Fixed, thanks!

gopiotr reviewed (changes requested):
Hmmmm, now I see another problem. `--build-only` and then `--test-only` sequence don't work correctly with this PR:
```
./scripts/twister -vv --inline-logs -p native_posix -T tests/kernel/poll --build-only
```
and then
```
./scripts/twister -vv --inline-logs -p native_posix -T tests/kernel/poll --test-only
```
Such sequence end with status "failed", reason "RunID mismatch". And this is because when you call `apply_filters` in second command (with `--test-only`), then new `TestInstance` objects are generated with new `run_id`. Old `run_id` should be inherited.

PerMac commented:
@nashif Can you give it another look? I addressed all the issues found by you and Piotr and I explained the motivation behind the changes.

nashif commented:
@PerMac  I am fine with making the skip reason clearer, but changing the workflow and applying filter on every call is not related and can have very serious consequences. So it need to come as a seperate PR after putting more thought. I do not think we should be changing the workflow and how we filter tests as a side effect of what seems like a harmless enhancement of making skip reason clearer.
